algorithms/symmetry/cosym/target.py

```diff
@@ -371,32 +366,13 @@ def compute_gradients(self, x: np.ndarray) -> np.ndarray:
           f: The value of the target function at coordinates `x`.
           grad: The gradients of the target function with respect to the parameters.
         """
-        grad = np.empty(x.shape)
+        x = x.reshape((self.dim, x.size // self.dim))
         if self.wij_matrix is not None:
             wrij_matrix = np.multiply(self.wij_matrix, self.rij_matrix)
+            grad = -2 * (x @ wrij_matrix - x @ np.multiply(self.wij_matrix, x.T @ x))
```

Comment:
Would this work?  Reducing the number of matrix multpilications simplifies the calculation.
```suggestion
            grad = -2 * x @ (wrij_matrix - np.multiply(self.wij_matrix, x.T @ x))
```

For comparison, a dumb example:
```
In [1]: import numpy as np

In [2]: x = np.random.randint(10, size=(2, 5))

In [3]: y = np.random.randint(10, size=(5, 2))

In [4]: z = np.random.randint(10, size=(5, 2))

In [5]: x, y, z
Out[5]: 
(array([[3, 0, 4, 0, 8],
        [1, 2, 8, 1, 0]]),
 array([[3, 9],
        [9, 7],
        [6, 9],
        [6, 8],
        [2, 2]]),
 array([[1, 6],
        [4, 3],
        [8, 6],
        [8, 3],
        [1, 0]]))

In [6]: %timeit x @ y - x @ z
2.22 µs ± 29.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit x @ (y - z)
1.33 µs ± 22.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
Thanks, good spot. Based on limited testing it doesn't appear to give a _significant_ performance benefit in the wider context, but it seems a sensible and useful improvement nonetheless 👍